File: backend/src/models/CostRecord.ts
import mongoose, { Schema, Document } from 'mongoose';
import { BaseModelPlugin } from './BaseModelPlugin';
import crypto from 'crypto';
import { ICloudResource } from './CloudResource';
import { ICloudProvider } from './CloudProvider';

export interface ICostRecord extends Document<any> {
    _id: Buffer | string;
    resource_id: string;
    provider_id: string;
    billing_period: string;
    daily_cost: number;
    weekly_cost: number;
    monthly_cost: number;
    projected_monthly_cost: number;
    currency: string;
    billing_status: string;
    cost_timestamp: Date;
    created_at: Date;
    updated_at: Date;
}

const costRecordSchema = new Schema<ICostRecord>({
    _id: { type: 'UUID', default: () => crypto.randomUUID() },
    resource_id: { type: String, ref: 'CloudResource', required: true },
    provider_id: { type: String, ref: 'CloudProvider', required: true },
    billing_period: { type: String, required: true, index: true },
    daily_cost: { type: Number, default: 0.0 },
    weekly_cost: { type: Number, default: 0.0 },
    monthly_cost: { type: Number, default: 0.0 },
    projected_monthly_cost: { type: Number, default: 0.0 },
    currency: { type: String, default: 'USD', index: true },
    billing_status: { type: String, default: 'pending', index: true },
    cost_timestamp: { type: Date, default: Date.now, index: true }
}, { collection: 'cost_records' });

costRecordSchema.plugin(BaseModelPlugin);

export const CostRecord = mongoose.model<ICostRecord>('CostRecord', costRecordSchema);

File: backend/app/models/cost_record.py
from datetime import datetime
from pydantic import Field
from beanie import Indexed, Link
from app.models.base import TimestampedDocument
from app.models.cloud_resource import CloudResource
from app.models.cloud_provider import CloudProvider
from app.models.base import get_utc_now

class CostRecord(TimestampedDocument):
    resource: Link[CloudResource]
    provider: Link[CloudProvider]
    billing_period: Indexed(str)
    daily_cost: float = Field(0.0, description="Daily cost in currency")
    weekly_cost: float = Field(0.0, description="Weekly cost in currency")
    monthly_cost: float = Field(0.0, description="Monthly cost in currency")
    projected_monthly_cost: float = Field(0.0, description="Projected monthly cost")
    currency: str = "USD"
    billing_status: Indexed(str) = "pending"
    cost_timestamp: Indexed(datetime) = Field(default_factory=get_utc_now)

    class Settings:
        name = "cost_records"
        indexes = [
            "billing_period",
            "currency",
            "billing_status",
            "cost_timestamp"
        ]
